cluster-autoscaler/debuggingsnapshot/debugging_snapshotter.go

```diff
@@ -0,0 +1,222 @@
+/*
+Copyright 2021 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package debuggingsnapshot
+
+import (
+	"net/http"
+	"sync"
+	"time"
+
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/klog/v2"
+	"k8s.io/kubernetes/pkg/scheduler/framework"
+)
+
+// DebuggingSnapshotterState is the type for the debugging snapshot State machine
+// The states guide the workflow of the snapshot.
+type DebuggingSnapshotterState int
+
+// DebuggingSnapshotterState help navigate the different workflows of the snapshot capture.
+const (
+	// SNAPSHOTTER_DISABLED is when debuggingSnapshot is disabled on the cluster and no action can be taken
+	SNAPSHOTTER_DISABLED DebuggingSnapshotterState = iota + 1
+	// LISTENING is set when snapshotter is enabled on the cluster and is ready to listen to a
+	// snapshot request. Used by ResponseHandler to wait on to listen to request
+	LISTENING
+	// TRIGGER_ENABLED is set by ResponseHandler if a valid snapshot request is received
+	// it states that a snapshot request needs to be processed
+	TRIGGER_ENABLED
+	// START_DATA_COLLECTION is used to synchronise the collection of data.
+	// Since the trigger is an asynchronous process, data collection could be started mid-loop
+	// leading to incomplete data. So setter methods wait for START_DATA_COLLECTION before collecting data
+	// which is set at the start of the next loop after receiving the trigger
+	START_DATA_COLLECTION
+	// DATA_COLLECTED is set by setter func (also used by setter func for data collection)
+	// This is set to let Flush know that at least some data collected and there isn't
+	// an error State leading to no data collection
+	DATA_COLLECTED
+)
+
+// DebuggingSnapshotterImpl is the impl for DebuggingSnapshotter
+type DebuggingSnapshotterImpl struct {
+	// State captures the internal state of the snapshotter
+	State *DebuggingSnapshotterState
+	// DebuggingSnapshot is the data bean for the snapshot
+	DebuggingSnapshot DebuggingSnapshot
+	// Mutex is the synchronisation used to the methods/states in the critical section
+	Mutex *sync.Mutex
+	// FlushTrigger is the channel on which the Response Handler waits on to know
+	// when there is data to be flushed back to the channel from the Snapshot
+	FlushTrigger chan bool
```

Comment:
I think selector would be more readable (and would allow you to avoid a bug I just mentioned in another comment). If you want to pass states through this channel though, an enum would be easier to follow.